bionic/code_references.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2018-2020 Streamlit Inc.
+# Author: Dominik Moritz
+
+"""
+Contains the logic to find all the referenced objects for a code object.
+"""
+
+import attr
+import dis
+import importlib
+import inspect
+import warnings
+
+from .utils.misc import oneline
+
+
+@attr.s(frozen=True)
+class CodeContext:
+    globals = attr.ib()
+    cells = attr.ib()
+    varnames = attr.ib()
+
+
+def get_code_context(func) -> CodeContext:
+    code = func.__code__
+
+    # Mapping from variable name to the value if we can resolve it.
+    # Otherwise map to the name.
+    cells = {}
+
+    for var in code.co_cellvars:
+        cells[var] = var  # Instead of value, we use the name.
+
+    if code.co_freevars:
+        assert len(code.co_freevars) == len(func.__closure__)
+        cells.update(
+            zip(code.co_freevars, map(lambda c: c.cell_contents, func.__closure__))
+        )
+
+    varnames = {}
+    if inspect.ismethod(func):
+        varnames = {"self": func.__self__}
+
+    return CodeContext(globals=func.__globals__, cells=cells, varnames=varnames)
+
+
+def get_referenced_objects(code, context):
+    # Top of the stack
+    tos = None
+    lineno = None
+    refs = []
+
+    def set_tos(t):
+        nonlocal tos
+        if tos is not None:
+            # Hash tos so we support reading multiple objects
+            refs.append(tos)
+        tos = t
+
+    # Our goal is to find referenced objects. The problem is that co_names
+    # does not have full qualified names in it. So if you access `foo.bar`,
+    # co_names has `foo` and `bar` in it but it doesn't tell us that the
+    # code reads `bar` of `foo`. We are going over the bytecode to resolve
```

Comment:
Do we have a test case that illustrates this `foo.bar` case and how it's useful? I see the example with `[MyClass, "log_val"]`, but it's not clear to me why we care about `"log_val"`, since presumably we can get that just from hashing the bytecode.